source/client-side-operations-timeout/client-side-operations-timeout.rst

```diff
@@ -0,0 +1,871 @@
+
+      #. MONGODB-X509: Execution of the commands required for the
+         authentication conversation.
+
+#. Client-side encryption
+
+   #. Execution of ``listCollections`` commands to get collection schemas.
+
+   #. Execution of ``find`` commands against the key vault collection to get
+      encrypted data keys.
+
+   #. Requests to non-local key management servers (e.g. AWS KMS) to decrypt
+      data keys.
+
+   #. Requests to mongocryptd servers.
+
+#. Socket write to send a command to the server
+
+#. Socket read to receive the server’s response
+
+The ``timeoutMS`` option MUST apply to all blocking sections. Drivers MUST
+document any exceptions. For example, drivers that do not have full control
+over OCSP verification might not be able to set timeouts for HTTP requests to
+responders and would document that OCSP verification could result in an
+execution time greater than ``timeoutMS``.
+
+Server Selection
+~~~~~~~~~~~~~~~~
+
+If ``timeoutMS`` is set, drivers MUST use ``min(serverSelectionTimeoutMS,
+remaining timeoutMS)``, referred to as ``computedServerSelectionTimeout`` as
+the timeout for server selection and connection checkout. The server selection
+loop MUST fail with a timeout error once the timeout expires.
+
+After a server has been selected, drivers MUST use the remaining
+``computedServerSelectionTimeout`` value as the timeout for connection
+checkout. If a new connection is required, ``min(connectTimeoutMS, remaining
+computedServerSelectionTimeout)`` MUST be used as the timeout for TCP socket
+establishment. Any network requests required to create or authenticate a
+connection (e.g. HTTP requests to OCSP responders) MUST use
+``min(operationTimeout, remaining computedServerSelectionTimeout)`` as a
+timeout, where ``operationTimeout`` is the specified default timeout for the
+network request. If there is no specified default, these operations MUST use
+the remaining ``computedServerSelectionTimeout`` value. All commands sent
+during the connection’s handshake MUST use the remaining
+``computedServerSelectionTimeout`` as their ``timeoutMS`` value. Handshake
+commands MUST also set timeouts per the `Command Execution`_ section.
+
+If ``timeoutMS`` is not set and support for ``waitQueueTimeoutMS`` has not
+been removed, drivers MUST continue to exhibit the existing timeout behavior
+by honoring ``serverSelectionTimeoutMS`` for server selection and
+``waitQueueTimeoutMS`` for connection checkout. If a new connection is
+required, drivers MUST use ``connectTimeoutMS`` as the timeout for socket
+establishment and ``socketTimeoutMS`` as the socket timeout for all handshake
+commands.
+
+See `serverSelectionTimeoutMS is not deprecated`_ and `connectTimeoutMS is
+not deprecated`_.
+
+Command Execution
+~~~~~~~~~~~~~~~~~
+
+If ``timeoutMS`` is set, drivers MUST append a ``maxTimeMS`` field to
+commands executed against a MongoDB server. If the 90th percentile RTT of the
```

Comment:
Is 90th percentile RTT the value that drivers already track?

Reply:
If not, we should be more prescriptive of the algorithm rather than having everyone research it themselves.

Reply:
No, it's not. I looked into this, but different languages will have varying libraries for this that implement different algorithms. I'm not sure how prescriptive we want to be here for that reason.

Reply:
Still confused.  I understand that there are different libraries, but don't all drivers have to have the same behavior?  Is "90th percentile RTT" precise enough?

Reply:
I made the following changes in my latest commit:

1. Added a `ninetiethPercentileRoundTripTime` to the ServerDescription type in the SDAM spec.
2. Moved the information about calculating a server's 90th percentile RTT using the t-digest algorithm to the Server Monitoring spec.
3. Modified the RTT thread pseudocode in the Server Monitoring spec to reflect the changes from (2).

Let me know if this addresses your concerns.